gloo/transport/tcp/helpers.h

```diff
@@ -0,0 +1,171 @@
+    loop_->registerDescriptor(socket_->fd(), EPOLLIN | EPOLLONESHOT, this);
+  }
+
+  void handleEvents(int events) override {
+    // Move leaked shared_ptr to the stack so that this object
+    // destroys itself once this function returns.
+    auto self = std::move(this->leak_);
+
+    // Read T.
+    auto rv = socket_->read(&t_, sizeof(t_));
+    if (rv == -1) {
+      fn_(socket_, SystemError("read", errno), std::move(t_));
+      return;
+    }
+
+    // Check for short read (assume we can read in a single call).
+    if (rv < sizeof(t_)) {
+      fn_(socket_, ShortReadError(rv, sizeof(t_)), std::move(t_));
+      return;
+    }
+
+    fn_(socket_, Error::kSuccess, std::move(t_));
+  }
+
+ private:
+  std::shared_ptr<Loop> loop_;
+  std::shared_ptr<Socket> socket_;
+  callback_t fn_;
+  std::shared_ptr<ReadValueOperation<T>> leak_;
+
+  T t_;
+};
+
+template <typename T>
+void read(
+    std::shared_ptr<Loop> loop,
+    std::shared_ptr<Socket> socket,
+    typename ReadValueOperation<T>::callback_t fn) {
+  auto x = std::make_shared<ReadValueOperation<T>>(
+      std::move(loop), std::move(socket), std::move(fn));
+  x->run();
+}
+
+// WriteValueOperation asynchronously writes a value of type T to the
+// socket specified at construction. Upon completion or error, the
+// callback is called. Its lifetime is coupled with completion of the
+// operation, so the called doesn't need to hold on to the instance.
+// It does so by storing a shared_ptr to itself (effectively a leak)
+// until the event loop calls back.
+template <typename T>
+class WriteValueOperation final
+    : public Handler,
+      public std::enable_shared_from_this<WriteValueOperation<T>> {
+ public:
+  using callback_t =
+      std::function<void(std::shared_ptr<Socket>, const Error& error)>;
+
+  WriteValueOperation(
+      std::shared_ptr<Loop> loop,
+      std::shared_ptr<Socket> socket,
+      T t,
+      callback_t fn)
+      : loop_(std::move(loop)),
+        socket_(std::move(socket)),
+        fn_(std::move(fn)),
+        t_(std::move(t)) {}
+
+  void run() {
+    // Cannot initialize leak until after the object has been
+    // constructed, because the std::make_shared initialization
+    // doesn't run after construction of the underlying object.
+    leak_ = this->shared_from_this();
+    // Register with loop only after we've leaked the shared_ptr,
+    // because we unleak it when the event loop thread calls.
+    loop_->registerDescriptor(socket_->fd(), EPOLLOUT | EPOLLONESHOT, this);
+  }
+
+  void handleEvents(int events) override {
+    // Move leaked shared_ptr to the stack so that this object
+    // destroys itself once this function returns.
+    auto leak = std::move(this->leak_);
+
+    // Write T.
+    auto rv = socket_->write(&t_, sizeof(t_));
+    if (rv == -1) {
+      fn_(socket_, SystemError("write", errno));
+      return;
+    }
+
+    // Check for short read (assume we can write in a single call).
```

Comment:
You mean a short write/ ShortWriteError?

I don't know the typical sizeof(t_) or what's typical here, but does it make sense to not accept a short write here, and instead schedule another write?

If the process is keeping the kernel buffers full, it might be common to have short writes, and isn't necessarily such a bad situation, unless you need to guarantee that sizeof(t_) goes out.

Reply:
Yes, good catch. This was a typical copy-paste error...

These helpers are used to send the sequence number around, which is only 8 bytes. Once that's done, the pair takes over control of the file descriptor and properly takes care of the partial reads/writes. Therefore I think it's fine to make this assumption here. The function is only called if `epoll(2)` says the socket is readable or writable, so we should be able to complete an 8 byte read/write.

Reply:
8 bytes should be quite benign then - I just didn't have context on the expected size here, wasn't sure if it was std::array<uint8_t, SomeBigN> or something :)